### PR TITLE
undercurls (fixes #136)

### LINKED ISSUE #136: add undercurls support
Undercurls are great because they are independent from the characters underneath them, which makes them very useful for hinting hints, warns, errs ..etc

This was originally a [kitty only](https://sw.kovidgoyal.net/kitty/underlines/) feature and then it was standardized as a specification and now most other terminals do support it.
Luckily this was already implemented into st as a [patch](https://st.suckless.org/patches/undercurl/) by HexOctal

here's how it looks like for both kitty and xst right now

### kitty
image

### xst
image

### PR DESCRIPTION
**Pull request by RaafatTurki**
See #136 for more info on what undercurls are. I've patched everything except for one conflict, xst compiles fine, however I still get underlines instead of undercurls.

That one conflict was [this line](https://gist.github.com/RaafatTurki/a32d8b5636fa7e2e17bc1505e0bb3d4b#file-st-undercurl-0-8-4-20210822-diff-L250) which removes an [XftDrawRect](https://github.com/gnotclub/xst/blob/d35241209e47ea40ba0ffdc63b04b7134295b76d/x.c#L1668) function call and replaces it with the actual implementation of the undercurls + a fallback to underlines

Here's a comparison of that function call 
### patch
```c
XftDrawRect(xw.draw, fg, winx, winy + dc.font.ascent + 1, width, 1);
```
### xst
```c
XftDrawRect(xw.draw, fg, winx, winy + win.cyo + dc.font.ascent + 1, width, 1);
```

They only differ in the addition of `win.cyo`, I have no idea what xst patch introduced this parameter, its significance nor how it should be used within the newly patched code.

Any hint would be highly appreciated.

### CODE REVIEW

**Comment by actionless** [comment]
> They only differ in the addition of win.cyo, I have no idea what xst patch introduced this parameter, it's significance nor how it should be used within the newly patched code.

> Any hint would be highly appreciated.

```console
$ rg cyo
x.c
88:     int cyo; /* char y offset */
1049:   win.cyo = ceilf(dc.font.height * (chscale - 1) / 2);
1319:                   yp = winy + font->ascent + win.cyo;
1565:           XftDrawRect(xw.draw, fg, winx, winy + win.cyo + dc.font.ascent + 1,
1570:           XftDrawRect(xw.draw, fg, winx, winy + win.cyo + 2 * dc.font.ascent / 3,
```

**Comment by RaafatTurki** [comment]
I think the issue is much larger than that, I've tried:
-  the patch on vanilla st
- a maintained [st fork](https://github.com/BeyondMagic/st) that claims to have undercurl support.

on 2 of my machines, no undercurls.

This is a major roadblock, could I be doing this all wrong? anyone who does the steps above let me know what you get.

**Comment by RaafatTurki** [comment]
terminfo is not fun

**Comment by actionless** [comment]
i could help reviewing the PR when it's ready but as i'm not interested in the feature myself i can't help with developing it, sorry

**Comment by RaafatTurki** [comment]
oh no worried, I have this in the bag ... sorta
image

**Comment by RaafatTurki** [comment]
I've got it working now but it looks a bit different due to how st renders characters
#### kitty
image
#### xst
image

it seems like st doesn't provide an extra padding around its characters so stuff like a undercurl could fit within its bounding box.
should I add the xrdb support and finalize this PR or is there an issue with the above?

File: x.c
Comment (actionless) [inline x.c:1721]: i believe whole this chunk is better to extract into separate function or even file

File: x.c
Comment (actionless) [inline x.c:1764]: and also you missed that `cyo` which you were asking about in one of previous comments

File: x.c
Comment (RaafatTurki) [reply x.c]: I've kept only one method to render undercurls so this should no longer be an issue, it also means xrdb support is no longer needed

**Comment by RaafatTurki** [comment]
I've selected the most wide spread (and sane) method of rendering undercurls, that has made them look nicer and not going above the characters baseline, it also scales with the font size.
#### 10px
image
#### 13px
image
#### 16px
image

**Comment by actionless** [comment]
you have to apply that `cyo` y axis offset to line itself as well, i guess

File: x.c
Comment (actionless) [reply x.c]: i'd still would wait for his feedback on this point before merging

Comment (neeasade) [reply x.c]: > i believe whole this chunk is better to extract into separate function or even file (but need to have @neeasade 's input on if we are trying to merge the st patches 'as is' or with reasonable codestyle improvements)

Prefering 'as is' when possible/separate files are better when feasible

File: st.info
Comment (neeasade) [inline st.info:2]: nitpick -- whitespace should match other lines/tabs (same thing with the `int ustyle;` declaration, and `int carg`)

File: x.c
Comment (RaafatTurki) [reply x.c]: @neeasade the patch includes 3 directive `#if` caluses controlled by a `#define` variable, 2 of these caluses render the undercurl in non-standard (and in a way that covers characters) methods, what I did was remove those and use the single clause which is now a normal `if` statement (check the latest commit)

I've also tried abstracting it into its own separate file but it's not doable since it's using data types defined within x.c itself (`XWindow` for example)

Comment (neeasade) [reply x.c]: Yes, that makes sense, thanks!. I think @actionless's suggestion to shove this into it's own function at least still would still be an improvement, as it's a large chunk dedicated to squiglies

Comment (RaafatTurki) [reply x.c]: sounds good. on a side note, the original patch does stuff like heavy commenting (even drawing ascii stuff) and breaking simple lines into multiple just for the sake of commenting them, take a look at this for example
image

does all that go into the extracted function?

Comment (actionless) [reply x.c]: it's a bit funky but not criminal :)

Comment (neeasade) [reply x.c]: > does all that go into the extracted function?

yes, I think that's fine/sounds good to me. I'm eyeing [this chunkish](https://github.com/gnotclub/xst/pull/137/files#diff-847ba343b012c9d0a16320307daf487df677d6945af4bdb1fb2b762f02def4daR1723-R1921)

Comment (actionless) [reply x.c]: 2021-12-02--1638403076_1595x625_scrot

File: x.c
Comment (actionless) [inline x.c:1888]: i think we could still factor all this out into one more "drawunderline" function, or just combine those two functions curls/underline into one

Comment (RaafatTurki) [reply x.c]: While that's valid I think it would be a bad idea to have such function that is essentially a
```c
void drawunderlineorundercurl() { ... }
```
since it will needs so much unrelated stuff passed into it but needed because a single property of an entire struct is needed.
a less severe refactor that follows the code logic could be the correct move, take a look at the latest commit as the whole undercurl logic is 11 lines now.

**Review by actionless — APPROVED** [review]

**Comment by actionless** [comment]
thanks! i think now it looks good to go

**Comment by RaafatTurki** [comment]
I went this morning and implemented the rest of the patch, the `cyo` was the cause of the rendering issue.
figured it would make the life of whoever comes after me easier plus more options so why not, xrdb support incoming.
#### curly
image
#### spiky
image
#### curly
image

**Review by actionless — APPROVED** [review]

**Comment by actionless** [comment]
are the latest screenshots made already with applying `cyo`?

after zooming them in they still dont look as good vertically adjusted as in kitty:

your last screenshot:

2021-12-02--1638460630_608x158_scrot


your kitty screenshot:

2021-12-02--1638460642_571x127_scrot


feels like it need to add(remove) 1 extra pixel of vertical coordinates

**Comment by actionless** [comment]
also, could you make a screenshot of text with with underline? (`_`) using the same font as you used for kitty

**Comment by actionless** [comment]
and, (here i am being like really picky) - because of using ceil instead of round - you have a rounding mistake in your formula resulting in the "tilt" of the sinewave right in the vertical middle of it:
2021-12-02--1638460969_726x43_scrot

**Comment by RaafatTurki** [comment]
the ceil has no effect on the wave tilt, however I've edited
```c
# x.c:1596
int wh = win.cyo + dc.font.descent - wlw/2 - 1;
to 
int wh = win.cyo + dc.font.descent - wlw/2 - 2;
```
and now it renders better
#### kitty
image
### xst
image

**Comment by actionless** [comment]
it must be somewhere the round() missing, i'll try to look through the code again to guess where

**Comment by actionless** [comment]
regarding vertical alignment - can confirm it looks uniform with kitty now :+1:

File: x.c
Comment (actionless) [inline x.c:1626]: you sure you've tried to replace both ceils to rounds here...

File: x.c
Comment (actionless) [inline x.c:1630]: ...and here?

File: x.c
Comment (actionless) [inline x.c:1608]: and mb round it here and next

File: x.c
Comment (actionless) [inline x.c:1606]: and mb round it here and next

File: x.c
Comment (actionless) [inline x.c:1614]: this one i meant as "next"

File: x.c
Comment (actionless) [reply x.c:1626]: nope, actually for first and last `ceil` makes sense, to avoid 'overflowing' from the width

**Comment by RaafatTurki** [comment]
The tilt seems to be directly effected by the font size which makes sense, the patch doesn't implement a sophisticated method for mapping pixel patterns to always represent a smooth curve. here are some samples:
#### 9px
image
#### 14px
image
#### 15px
image
#### 16px
image
#### 17px
image

This issue is mostly apparent in 15 and 16 but it does show in larger sizes, the gist of it is the pixel plotting works flawlessly on anything less than 15 because it's so restricted the "sin wave" always becomes a square wave, but anything higher and you enter "is this font size going to produce a nice wave or is it going to leave a dangling pixel as a side effect" land.

This whole issue becomes irrelevant if you use any of the other 2 methods as they use a less float-dependant methods of rendering, which makes me think maybe I should make the default one of them?

**Comment by actionless** [comment]
it seems what just on those 2 font sizes the values getting their 'edge case' to make the rounding problem visible

but since i dont have personal interest in this thing - that won't be a blocker from merging it from my side, so there is no urge in fixing it now

**Comment by RaafatTurki** [comment]
cool, I have nothing more to add to this PR, I'll wait for the merge.

**Comment by RaafatTurki** [comment]
@neeasade review pls

**Comment by neeasade** [comment]
Looks good, thanks again

**Pull request merged**
